Support for elevation based dark theme overlay color in the Material widget

darrenaustin opened the pull request:
## Description

In dark themes, drop shadows used to indicate elevation in Material apps can be difficult to see. To help with this, the [Material Dark theme specification](https://material.io/design/color/dark-theme.html#properties), adds a new semi-transparent overlay on surfaces to lighten them according to their elevation. This PR adds support for this feature.

![mio-design_assets_1A1Uf3a9TsbtgSKT-hjIAaEiEIXFDTUfE_darktheme-elevationsystem](https://user-images.githubusercontent.com/19588/60552610-53910400-9ce4-11e9-880d-33c221bf8204.png)

If a `Material` widget is in a theme where the new `ThemeData.applyElevationOverlay` flag is turned on, and the `Material.color` is the theme's surface color (i.e. `ThemeData.colorScheme.surface`), then a semi-transparent white will be composited over the `Material.color` for the background of the widget. The level of the transparency follows the values in the specification.

By default this feature is turned off to maintain backwards compatibility with apps built before the Dark theme spec was published. 

To turn it on for your app, you can adjust your dark themes to have `ThemeData.applyElevationOverlay` set to `true`.

## Related Issues

#35494 - A previous attempt at this PR.

## Tests

I added a group of new tests in material_test.dart to exercise this feature.

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] No, this is not a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

darrenaustin committed:
Added support for dark theme overlay colors to the `Material` widget. Also
added a `darkThemeOverlay` boolean to turn it off in cases where it not
wanted.

darrenaustin committed:
Added applyDarkThemeElevationOverlay to ThemeData:

- Renamed Material.darkThemeOverlay -> Material.applyDarkThemeElevationOverlay
- Added applyDarkThemeElevationOverlay to ThemeData
- Modified Material to use the theme's overlay setting if none is provided
  to the widget.
- Turned the overlay off by default

HansMuller commented on packages/flutter/lib/src/material/material.dart:
```diff
@@ -200,15 +201,17 @@ class Material extends StatefulWidget {
-  /// Defaults to 0. Changing this value will cause the shadow to animate over
-  /// [animationDuration].
+  /// Defaults to 0. Changing this value will cause the shadow and dark theme
+  /// overlay to animate over [animationDuration].
```
and dark => and the

I wish there was a link to an explanation of "dark theme overlay" in here somewhere. Maybe insert a `[applyDarkThemeOverlay]` somewhere.

HansMuller commented on packages/flutter/lib/src/material/material.dart:
```diff
@@ -217,9 +220,31 @@ class Material extends StatefulWidget {
   ///
+  /// If inside a dark theme and [applyDarkThemeElevationOverlay] is [true],
+  /// a semi-transparent white will be composited on top this color to
+  /// indicated the elevation.
```
indicated => indicate

HansMuller commented on packages/flutter/lib/src/material/material.dart:
```diff
@@ -217,9 +220,31 @@ class Material extends StatefulWidget {
+  ///
+  /// If [false] the [color] will be used unmodified.
+  ///
+  /// If [null] then it will use the value of the surrounding
```
It wouldn't hurt to say explain the need for this - because shadows are hard to see in the dark - in this doc.

HansMuller commented on packages/flutter/lib/src/material/material.dart:
```diff
@@ -217,9 +220,31 @@ class Material extends StatefulWidget {
+  /// [ThemeData.applyDarkThemeElevationOverlay] which defaults to [false].
+  ///
+  /// See also:
+  ///   * [Material.color]
```
Usually we provide a brief explanation for each see also.  The explanation makes it clear why the linked-to thing is relevant.

HansMuller commented on packages/flutter/lib/src/material/material.dart:
```diff
@@ -200,15 +201,17 @@ class Material extends StatefulWidget {
   ///
-  /// This controls the size of the shadow below the material.
+  /// This controls the size of the shadow below the material and the dark
+  /// theme overlay color if inside a dark theme.
```
We're saying "dark theme" quite a bit, however from the API's point of view it's a theme for which `brightness` is `Brightness.dark`. It would be good to draw attention to that somewhere.

HansMuller commented on packages/flutter/lib/src/material/material.dart:
```diff
@@ -305,16 +331,44 @@ class _MaterialState extends State<Material> with TickerProviderStateMixin {
-      default:
-        return null;
+    Color color = widget.color;
+    if (color == null) {
```
Usually we create a temporary theme variable like:
`final ThemeData theme = Theme.of(context);`

HansMuller commented on packages/flutter/lib/src/material/material.dart:
```diff
@@ -305,16 +331,44 @@ class _MaterialState extends State<Material> with TickerProviderStateMixin {
-        return null;
+    Color color = widget.color;
+    if (color == null) {
+      if (widget.type == MaterialType.canvas) {
```
We always `switch` enums: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-if-chains-or--or--with-enum-values

HansMuller commented on packages/flutter/lib/src/material/theme_data.dart:
```diff
@@ -679,6 +683,28 @@ class ThemeData extends Diagnosticable {
+  /// If [false] the surface color will be used unmodified.
+  ///
+  /// Defaults to [false].
+  ///
```
Same comment as earlier about explaining the need for this (dark shadows, etc).

It looks like there are some opportunities for using `@template` and `@macro` here.

HansMuller reviewed (approved):
LGTM, mostly just some comments about the docs.

rami-a reviewed (approved):
LGTM!

darrenaustin committed:
Updated from review comments

- Removed applyDarkThemeElevationOverlay from Material as it is covered
  by ThemeData.
- Updated documentation based on feedback.
- Made just the overlay color animate to get around issues with
  OutlineButton.

darrenaustin committed:
Fixed an indent problem.

darrenaustin commented:
> The indentation looks off here

D'oh. Old habits die hard. Thanks.

darrenaustin committed:
More updates from feedback:
- Only apply the overlays to Material widgets using the surface color.
- Switched to using the elevation->opacity formula that Android is using.
- Renamed applyDarkThemeElevationOverlay -> applyElevationOverlay.
- No longer checking against ThemeData.brightness to dark as
  applyElevationOverlay should only be set to true on dark themes.

darrenaustin committed:
Minor comment fixes.

darrenaustin committed:
Minor comment fixes.

darrenaustin committed:
Fixed a formatting issue.

darrenaustin committed:
Fixed some missing const constructors.

rami-a commented on packages/flutter/test/material/material_test.dart:
```diff
@@ -208,6 +216,77 @@ void main() {
+          ),
+          child: buildMaterial(color: surfaceColor, elevation: 8.0))
+      );
+      final RenderPhysicalShape model = getShadow(tester);
```
Is there a reason the helper function is called `getShadow`? I was a little confused by this

rami-a reviewed (approved):
LGTM!

HansMuller commented on packages/flutter/lib/src/material/material.dart:
```diff
@@ -301,20 +316,43 @@ class Material extends StatefulWidget {
+// indicate the level of elevation.
+Color _elevationOverlayColor(BuildContext context, Color background, double elevation) {
+  final ThemeData theme = Theme.of(context);
+  if (elevation > 0.0 &&
```
Conventional formatting for this would be one-line or:
```
  if (elevation > 0.0 &&
      theme.applyElevationOverlay &&
      background == theme.colorScheme.surface) {
```

HansMuller commented on packages/flutter/lib/src/material/theme_data.dart:
```diff
@@ -159,6 +159,7 @@ class ThemeData extends Diagnosticable {
     ChipThemeData chipTheme,
     TargetPlatform platform,
     MaterialTapTargetSize materialTapTargetSize,
+    bool applyElevationOverlay,
```
Maybe applyElevationOverlayColor?

HansMuller commented on packages/flutter/lib/src/material/theme_data.dart:
```diff
@@ -679,6 +683,38 @@ class ThemeData extends Diagnosticable {
+  ///
+  /// Material drop shadows can be difficult to see in a dark theme, so the
+  /// elevation of a surface should be portrayed with an "overlay" in addition
+  /// to the shadow. As the elevation of the component increases, the bright
```
the bright => the white

HansMuller commented on packages/flutter/lib/src/material/theme_data.dart:
```diff
@@ -679,6 +683,38 @@ class ThemeData extends Diagnosticable {
+  /// Note: this setting is here to maintain backwards compatibility with
+  /// apps that were built before the Material Dark theme specification
+  /// was published. New apps should set this to [true] for dark themes
+  /// (where [brightness] is [Brightness.dark].
```
`(where` => where

HansMuller commented on packages/flutter/lib/src/material/theme_data.dart:
```diff
@@ -679,6 +683,38 @@ class ThemeData extends Diagnosticable {
+  ///
+  /// See also:
+  ///
+  ///  * [Material.elevation], which effects how transparent the overlay is.
```
the overlay => the white overlay

HansMuller commented on packages/flutter/lib/src/material/theme_data.dart:
```diff
@@ -679,6 +683,38 @@ class ThemeData extends Diagnosticable {
+  /// See also:
+  ///
+  ///  * [Material.elevation], which effects how transparent the overlay is.
+  ///  * [Material.color], which if it matches [colorScheme.surface] an overlay
```
* [Material.color], the white color overlay will only be applied of the material's color is [colorScheme.surface].

darrenaustin replied:
Much clearer. Thx.

HansMuller commented on packages/flutter/test/material/material_test.dart:
```diff
@@ -208,6 +216,77 @@ void main() {
+              ),
+            )
+        );
+        await tester
```
this can be one line

HansMuller reviewed (approved):
LGTM

darrenaustin committed:
Updated with feedback from review

- Minor formatting and doc changes
- Renamed ThemeData.applyElevationOverlay -> ThemeData.applyElevationOverlayColor
